File: src/DronePhysics.cpp
#include "DronePhysics.h"
#include "cmath"

double DronePhysics::DegreesToRadians(double degrees) {
    return ( degrees * M_PI) / 180 ;;
}

Drone::State DronePhysics::Fdt(Drone &drone, double h) {

    //

    Drone::State dX;
    dX.position = pathDefinition(drone, h);

    return dX;
}

Drone DronePhysics::rungeKutta(Drone &drone, double &h) {

    Drone::State xdt = Fdt(drone, h);
    drone.state = drone.state + h * xdt;
    //std::cout << "AFTER: " << drone.state.position.x << " " << drone.state.position.y << std::endl;
    return drone;
}

Vector DronePhysics::pathDefinition(Drone &drone, double h) {
    // здесь перебираем вектора, выбираем наилучший

    double angle = this->DegreesToRadians(20);
    Vector v = drone.state.velocity.rotate(-angle);
    Vector pBest;
    double best = 0.0;
    for(int i = 0; i<3; i++){
        double current = quality(drone, v.rotate(i*angle), h);
        if(current > best){
            best = current;
            pBest = v;
        }
    }

    return pBest;
}

double DronePhysics::quality(const Drone &drone, Vector nextVelocity, double &h) {

   // Drone droneNext = droneCurrent;
    //droneNext.state.d

    // f = Ttarget + Tclose
    double Ttarget = (((drone.target - drone.state.position) //to target
                    - (drone.target - drone.state.position + nextVelocity * h))).length() / nextVelocity.length();

    double radius = 10;
    double nextClosest = drone.neighbors[0].state.position.distanceTo(drone.state.position + nextVelocity * h);
    double currentClosest = drone.neighbors[0].state.position.distanceTo(drone.state.position);
    for(auto& neighbour : drone.neighbors){
        double nextDist = neighbour.state.position.distanceTo(drone.state.position + nextVelocity * h);
        double currentDist = neighbour.state.position.distanceTo(drone.state.position + nextVelocity * h);
        if(nextDist < nextClosest){
            nextClosest = nextDist;
            currentClosest = currentDist;
        }
    }

    if(nextClosest < radius){
        return .0;
    }

    double Tclose = (currentClosest - nextClosest) / nextVelocity.length();

    std::cout << "F: " << std::endl;
    std::cout << "target = " << Ttarget << "\nclose = "<< Tclose << std::endl;

    return Ttarget + Tclose;
}

File: src2/DronePhysics.py
import math
from typing import Type

import copy
from Drone import Drone
from State import State
from Vector import Vector
from src2.State import State


class DronePhysics:

    def quality(self, drone: Drone, next_velocity: Vector, h: float) -> float:
        if drone.state.velocity.length() == 0:
            return .0

        h*=3

        # f = Ttarget + Tclose
        Ttarget = ((drone.target.distance_to(drone.state.position))
                    - (drone.target.distance_to(drone.state.position + next_velocity * h))) / (next_velocity * h).length()

        radius = 50
        next_closest = current_closest = radius * radius
        for neighbour in drone.neighbors:
            if drone.state.position.distance_to(neighbour.state.position) <= radius:
                nextDist = neighbour.state.position.distance_to(drone.state.position + next_velocity * h)
                currentDist = neighbour.state.position.distance_to(drone.state.position)
                if next_closest > nextDist:
                    next_closest = nextDist
                    current_closest = currentDist

        if next_closest == current_closest == radius * radius:
            Tclose = 0
        else:
            Tclose = (next_closest - current_closest) / (next_velocity*h).length()

        # if next_closest < radius:
        #     # TODO: problem
        #     # print(f"PROBLEM DRONE: {drone}")
        #     return .0



        # print(f"Drone {drone.id}: Ttarget({Ttarget}) and Tclose({Tclose})")
        # print(Tclose)

        return Ttarget + Tclose

    def pathDefinition(self, drone: Drone, h: float) -> Vector:
        # здесь перебираем вектора, выбираем наилучший
        angle = math.radians(20)

        v: Vector = Vector(drone.state.velocity.x, drone.state.velocity.y, drone.state.velocity.z)
        v.rotate_z(-5*angle)

        pBest = Vector(0, 0, 0)
        best = 0
        # print("_-__________________")
        for i in range(1, 19):
            current = self.quality(drone, v.rotate_z(angle), h)
            #print(f'hereeee -------------{v}')
            #print(current)
            if current > best:
                best = current
                pBest = Vector(v.x, v.y, v.z)
        #if best == 0 and pBest == Vector(0, 0, 0):

        drone.quality = round(best, 3)
        return pBest

    def Fdt(self, drone: Drone, h: float) -> State:
        dX = State()
        # print(f"Before: {drone.state.position}")
        # dX.velocity = self.pathDefinition(drone, h) - drone.state.velocity
        # print(dX.velocity) self.pathDefinition(drone, h)
        # dX.velocity =  - drone.state.velocity #self.pathDefinition(drone, h)
        dX.position = self.pathDefinition(drone, h) # - drone.state.velocity
        # dX.velocity = Vector(drone.state.velocity.x,0,0)
        # print(f"After: {dX.position}")
        # dX.position = drone.state.velocity
        return dX

    def rungeKutta(self, drone: Drone, h: float):
        xdt = self.Fdt(drone, h)
        b = drone.state.velocity.length()
        drone.state = drone.state + (h * xdt)
        a = drone.state.velocity.length()
        # print(f'change in speed{b - a}')
        # print(drone.state.velocity)
        return drone
